src/main/java/org/jenkinsci/plugins/github_branch_source/ApiRateLimitChecker.java

```diff
@@ -0,0 +1,179 @@
+package org.jenkinsci.plugins.github_branch_source;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.Util;
+import hudson.model.TaskListener;
+import org.kohsuke.github.GHRateLimit;
+import org.kohsuke.github.GitHub;
+
+import java.io.IOException;
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+
+public enum ApiRateLimitChecker {
+    /**
+     * Attempt to evenly distribute GitHub API requests.
+     */
+    ThrottleForNormalize(Messages.ApiRateLimitChecker_ThrottleForNormalize()) {
+        @Override
+        public void checkApiRateLimit(@NonNull TaskListener listener, GitHub github) throws IOException, InterruptedException {
```

Comment:
This is mostly refactoring from https://github.com/jenkinsci/github-branch-source-plugin/pull/242/files#diff-61948c4cd4e480f18c6e572cf88a055bR572

Reply:
Yes, that was intentional. @alexanderrtaylor and I noticed that when that when the reset time has past, the rateLimitResetMillis could be negative.  That could make the calculated `ideal` go negative. 
This could in an edge case of line 39 allow users to make calls even when `rateLimit.remaining` is 0.